include/sddf/gpu/queue.h

```diff
@@ -0,0 +1,286 @@
+    GPU_RESP_ERR_INVALID_BOUNDS,
+    /* Invalid parameter in request */
+    GPU_RESP_ERR_INVALID_PARAMETER,
+} gpu_resp_status_t;
+
+typedef struct gpu_req {
+    gpu_req_code_t code;
+    /* Request id */
+    uint32_t id;
+    union {
+        gpu_req_get_display_info_t get_display_info;
+        gpu_req_resource_create_blob_t resource_create_blob;
+        gpu_req_set_scanout_blob_t set_scanout_blob;
+        gpu_req_resource_create_2d_t resource_create_2d;
+        gpu_req_resource_attach_backing_t resource_attach_backing;
+        gpu_req_resource_detach_backing_t resource_detach_backing;
+        gpu_req_set_scanout_t set_scanout;
+        gpu_req_transfer_to_2d_t transfer_to_2d;
+        gpu_req_resource_flush_t resource_flush;
+        gpu_req_resource_unref_t resource_unref;
+    };
+} gpu_req_t;
+
+typedef struct gpu_resp {
+    gpu_resp_status_t status;
+    /* Stores corresponding request id */
+    uint32_t id;
+} gpu_resp_t;
+
+/* Circular buffer containing requests */
+typedef struct gpu_req_queue {
+    uint32_t head;
+    uint32_t tail;
+    gpu_req_t buffers[];
+} gpu_req_queue_t;
+
+/* Circular buffer containing responses */
+typedef struct gpu_resp_queue {
+    uint32_t head;
+    uint32_t tail;
+    gpu_resp_t buffers[];
+} gpu_resp_queue_t;
+
+typedef struct gpu_queue_handle {
+    gpu_req_queue_t *req_queue;
+    gpu_resp_queue_t *resp_queue;
+    uint32_t capacity;
+} gpu_queue_handle_t;
+
+/**
+ * Initialise the shared queues.
+ *
+ * @param h queue handle to use.
+ * @param request pointer to request queue in shared memory.
+ * @param response pointer to response queue in shared memory.
+ * @param capacity number of entries in the req and resp queues.
+ */
+static inline void gpu_queue_init(gpu_queue_handle_t *h, gpu_req_queue_t *request, gpu_resp_queue_t *response,
+                                  uint32_t capacity)
+{
+    h->req_queue = request;
+    h->resp_queue = response;
+    h->capacity = capacity;
+}
+
+/**
+ * Check if the request queue is empty.
+ *
+ * @param h queue handle containing request queue.
+ *
+ * @return true indicates the request queue is empty, false otherwise.
+ */
+static inline bool gpu_queue_empty_req(gpu_queue_handle_t *h)
+{
+    return h->req_queue->tail - h->req_queue->head == 0;
+}
+
+/**
+ * Check if the response queue is empty.
+ *
+ * @param h queue handle containing response queue.
+ *
+ * @return true indicates the response queue is empty, false otherwise.
+ */
+static inline bool gpu_queue_empty_resp(gpu_queue_handle_t *h)
+{
+    return h->resp_queue->tail - h->resp_queue->head == 0;
+}
+
+/**
+ * Check if the request queue is full.
+ *
+ * @param h queue handle containing request queue.
+ *
+ * @return true indicates the request queue is full, false otherwise.
+ */
+static inline bool gpu_queue_full_req(gpu_queue_handle_t *h)
+{
+    return h->req_queue->tail - h->req_queue->head + 1 == h->capacity;
```

Comment:
This `+ 1` is not necessary anymore right? We can follow what the networking queues etc do now.